docs: update ABS feedback from beta (#1257)
redpanda-data/documentation: modules/develop/pages/transactions.adoc

== Use transactions

> By default, the `enable_transactions` cluster configuration property is set to true. However, in the following use cases, clients must explicitly use the Transactions API to perform operations within a transaction:

Reviewer: consider linking to the prperty

Reviewer 2: Added

* xref:develop:transactions.adoc#atomic-publishing-of-multiple-messages[Atomic (all or nothing) publishing of multiple messages]
* xref:develop:transactions.adoc#exactly-once-stream-processing[Exactly-once stream processing]

> The required `transactional.id` property acts as a producer identity. It enables reliability semantics that span multiple producer sessions by allowing the client to guarantee that all transactions issued by the client with the same ID have completed prior to starting any new transactions.

Reviewer 2: Is `transactional.id` a property set by a client, so not a Redpanda config property?

Reviewer 3: correct.

=== Atomic publishing of multiple messages

> A banking IT system with an event-sourcing microservice architecture illustrates the necessity for transactions. A bank has multiple branches, and each branch is an independent microservice that manages its own non-intersecting set of accounts. Each branch keeps its own ledger, represented as a Redpanda partition. When a branch (microservice) starts, it replays its ledger to reconstruct the actual state.

Reviewer: this could be simplified. Some terms here might not be widely recognized, specially by non-native speakers. One thing that makes me question this is "ledger". It's unclear if it's a tech-term or bank-term.

Check if the following makes sense in this context:

```suggestion
A banking system with an event-sourcing microservice architecture illustrates the necessity for transactions. A bank has multiple branches, and each branch is an independent microservice that manages its own accounts. Each branch stores its transaction history in a Redpanda partition. When a branch starts, it replays this history to reconstruct the current account balances.
```

Reviewer 2: Updated

The transformation reads a record from `topic(1)`, processes it, and writes it to `topic(2)`. Without transactions, an intermittent error can cause a message to be lost or processed several times. With transactions, Redpanda guarantees exactly-once semantics. For example:

.Show exactly-once processing example:
[%collapsible]
====
[,java]
----
var source = "source-topic";
var target = "target-topic";

Properties pprops = new Properties();
pprops.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "...");
pprops.put(ProducerConfig.ACKS_CONFIG, "all");
pprops.put(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, true);

> pprops.put(ProducerConfig.TRANSACTIONAL_ID_CONFIG, UUID.randomUUID().toString());

Reviewer 2: This is a suggestion from an AI review, need confirmation that this is correct.

Reviewer 3: This looks fine.

    }
}
----
====

==== Exactly-once processing configuration requirements

Redpanda’s default configuration supports exactly-once processing. To preserve this capability, ensure the following settings are maintained:

* `enable_idempotence = true`
* `enable_transactions = true`

> * `transaction_coordinator_delete_retention_ms` is greater than or equal to `transactional_id_expiration_ms`

Reviewer 2: This is specified in the [transaction_coordinator_delete_retention_ms](https://deploy-preview-1257--redpanda-docs-preview.netlify.app/current/reference/properties/cluster-properties/#transaction_coordinator_delete_retention_ms), is it correct to call it out here? Is "To ensure transaction state is retained as long as the longest-running transaction" commonly recommended / best practice?

Reviewer 3: correct.

== Best practices

To help avoid common pitfalls and optimize performance, consider the following when configuring transactional workloads in Redpanda:

> * Ongoing transactions can prevent consumers from advancing. To avoid this, don't set transaction timeout (`transaction.timeout.ms` in Java client) to high values: the higher the timeout, the longer consumers may be blocked. By default, it's about a minute, but it's a client setting that depends on the client.

Reviewer 4: ```suggestion
* Ongoing transactions can prevent consumers from advancing. To avoid this, don't set transaction timeout (`transaction.timeout.ms` in Java client) to high values: the longer the timeout, the longer consumers may be blocked. By default, it's about a minute, but it's a client setting that depends on the client.
```

> * Ongoing transactions can prevent consumers from advancing. To avoid this, don't set transaction timeout (`transaction.timeout.ms` in Java client) to high values: the longer the timeout, the longer consumers may be blocked. By default, it's about a minute, but it's a client setting that depends on the client.

Reviewer: > Ongoing transactions can prevent consumers from advancing.

Advancing what? offsets?

Reviewer: also, why do we mention the java client specifically? This is a kafka setting, so it's likely that all supported clients have this value for their producers.

Reviewer: > it's about a minute,

This is vague. We could be precise here.

Reviewer: > it's a client setting that depends on the client.

Could be simplified.

Reviewer: ```suggestion
* Ongoing transactions can prevent consumers from advancing their offsets. To avoid this, don't set transaction timeout (`transaction.timeout.ms`) to high values: the longer the timeout, the longer consumers may be blocked. The default is 60,000 ms (60 seconds).
```

Reviewer 2: @bharathv would you mind confirming if the suggested changes are correct?

Reviewer 3: > Advancing what? offsets?

When a Kafka consumer is configured to use the read_committed isolation level, it will only process messages that have been part of successfully committed transactions. This mechanism ensures data consistency, preventing consumers from acting on incomplete or potentially aborted transactions. However, a stuck transaction can prevent this process from moving forward, impacting consumer progress and potentially causing a backlog. 

A stuck transaction with a large timeout can block other committed transactions, this is by design, so large timeouts and stuck transactions results in seemingly stuck consumers.

Reviewer 3: > also, why do we mention the java client specifically? This is a kafka setting, so it's likely that all supported clients have this value for their producers.

Just to be a clear that is a Kafka "client" setting. It is named `transaction.timeout.ms` in Kafka Java client implementation but could be named something else in a different implementation. I think it refers to Java client because it is the most popular one.

Reviewer 3: > it's about a minute,

Exactly a minute, 60000ms to be exact (since the configuration is in ms).

ifndef::env-cloud[]

> * When running transactional workloads from clients, tune xref:reference:cluster-properties#max_transactions_per_coordinator[`max_transactions_per_coordinator`] to the number of active transactions that you expect your clients to run at any given time (if your client transaction IDs are not reused). 

Reviewer: ```suggestion
* When running transactional workloads from clients, tune xref:reference:cluster-properties#max_transactions_per_coordinator[`max_transactions_per_coordinator`] to match the number of concurrent transactions your clients run (if your client transaction IDs are not reused).
```

"any given time" is somewhat vague. Replaced with concurrent.

+

> The total number of transactions in the cluster at any one time is `max_transactions_per_coordinator * transaction_coordinator_partitions` (`transaction_coordinator_partitions` default is 50). When the threshold is exceeded, Redpanda terminates old sessions. If an idle producer corresponding to the terminated session wakes up and produces, its batches are rejected with the message `invalid producer epoch` or `invalid_producer_id_mapping`, depending on where it is in the transaction execution phase.

Reviewer: ```suggestion
The cluster's total transaction limit is `max_transactions_per_coordinator * transaction_coordinator_partitions` (default is 50 partitions). When this limit is exceeded, Redpanda terminates old sessions. If a terminated producer later tries to produce data, Redpanda rejects its batches with `invalid producer epoch` or `invalid_producer_id_mapping` errors.
```

This phrase flow is kinda awkward. Here's a suggestion. Feel free to pick whatever works best.

Reviewer 2: Updated
